From 92264efdb987d85b471b3a73931b4066510b5c16 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Fri, 19 Oct 2018 12:06:18 +0200
Subject: [PATCH 3/3] Merge android-9.0.0_r12

commit 7259d40f8c86da71557619f3368ba8899fcf0d3e
Merge: 2a9e3ccb4 7249fcd01
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Fri Jul 20 23:46:21 2018 +0000

    Snap for 4905103 from 7249fcd01328be699490ee792b228e89ef403459 to pi-dr1-release

    Change-Id: I25530e928971ea6220f4815f69cff00e212cbef0

commit 7249fcd01328be699490ee792b228e89ef403459
Author: Bill Yi <byi@google.com>
Date:   Thu Jul 19 17:22:29 2018 -0700

    Import translations. DO NOT MERGE

    Auto-generated-cl: translation import

    Bug: 64712476
    Change-Id: If947c7497c2e447fcd1a1f164dbc24ffc2c7fb69

commit 2a9e3ccb4da369e1c090cd28dc5a9b9cb0e582a9
Merge: 8c7ea0d2a f590f6f9c
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Jul 18 03:07:03 2018 +0000

    Snap for 4899218 from f590f6f9cd9b646a906bd967af24d6f279b533ae to pi-dr1-release

    Change-Id: Ie7db50162a0a6a848fa384c00187092914c2eff0

commit f590f6f9cd9b646a906bd967af24d6f279b533ae
Author: Pavlin Radoslavov <pavlin@google.com>
Date:   Tue Jul 10 17:24:01 2018 -0700

    Explicitly mute the audio output while switching the A2DP Active Device

    The mute/unmute of the audio output is needed during the A2DP
    ActiveDevice switch to avoid audio glitches.

    Bug: 78152025
    Bug: 110441865
    Test: Manual - stream with Pandora/Spotify and switch A2DP Active Device
    Change-Id: I5a233460a07e0d4fffcb4e6cb208d846517f0c1e
    (cherry picked from commit 075005f0ef8adf5397b9ed6210e0160f1a0314a5)

commit 8c7ea0d2add3d99fc25ee1481f7746a982e72f09
Merge: 99c293068 4c0065249
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Sun Jul 1 03:08:27 2018 +0000

    Snap for 4872253 from 4c00652497c48d4ac573307764824d8b262a82b8 to pi-dr1-release

    Change-Id: I58c0953905a40f6cdef7ff550f3a0c3189ad9ec6

commit 4c00652497c48d4ac573307764824d8b262a82b8
Merge: f1fb26b59 bd3055fe9
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Thu Jun 28 07:54:48 2018 +0000

    Merge "Always assign a free ID to the new MediaPlayer" into pi-dev

commit f1fb26b5936d952a9a678e31a6ffe093ef9a2138
Merge: ad387b762 3bb8f7841
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Thu Jun 28 06:28:43 2018 +0000

    Merge "HFP: Use SignalStrength#getLevel() to get signal level" into pi-dev

commit ad387b76244e41de22ae32c56177ede5619ddc38
Merge: 2ce6cb36f 92641885b
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Thu Jun 28 05:52:17 2018 +0000

    Merge "SCO: Don't set active device if a diconnect message is defered" into pi-dev

commit 99c293068b94c5cd714788adb70408afe50dd231
Merge: 4d25ea631 4da6bfdfb
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Thu Jun 28 03:32:16 2018 +0000

    Merge "Snap for 4866883 from 2ce6cb36f1efe1adc86708e9f664594c179bc93b to pi-dr1-release" into pi-dr1-release

commit 4da6bfdfb8fcba86ca383f9b0136e8cda3c50049
Merge: 51a648ef8 2ce6cb36f
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Thu Jun 28 03:23:37 2018 +0000

    Snap for 4866883 from 2ce6cb36f1efe1adc86708e9f664594c179bc93b to pi-dr1-release

    Change-Id: I1dbb89b5b35b00ab6df61abae21157fca886c6d9

commit 4d25ea63118f2d6f4e109ba121af38919bfc8c5f
Merge: 51a648ef8 2ce6cb36f
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Thu Jun 28 03:17:04 2018 +0000

    Snap for 4866863 from 2ce6cb36f1efe1adc86708e9f664594c179bc93b to pi-dr1-release

    Change-Id: I16c9dc494c9af9b55e3d5779e036402a1fa0d7c6

commit 2ce6cb36f1efe1adc86708e9f664594c179bc93b
Author: Jack He <siyuanh@google.com>
Date:   Tue Jun 26 15:36:22 2018 -0700

    HFP: Suspend A2DP whenever call state is busy

    * Suspend A2DP whenever call state is ringing or ongoing
    * Resume A2DP when call state becomes idle and no SCO audio is
      established

    Bug: 110771727
    Test: Receive incoming call when in-band ringing is disabled
    Change-Id: I59361384c223e64e7ce73444c4e1cdf25bf18781
    (cherry picked from commit 1192088a974f2e3b514c8b12cbfb3bf574a228aa)

commit 92641885b10d3e29af0742a2f1ae18afd68dd837
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Jun 22 21:59:31 2018 +0800

    SCO: Don't set active device if a diconnect message is defered

    - Fix a pin-pon scenario in HeadsetService.
    - If a headset B is set as active deivce while headset A is connecting SCO,
      HeadsetService request headset A's state machine to do disconnect SCO.
    - However, when state machine is at AudioConnecting state, DISCONNECT_AUDIO
      would be pending in defered meesage queue and will be handled in next state.
    - Once headset A entered AndioOn state, it set it self to as active device again
      then start handling DISCONNECT_AUDIO.
    - After headset A finally disconnect SCO, HeadsetService won't reconnect
      it and rout to headset B since active device is still headset A.
    - Active deivce would be like A->B->A, and no device SCO stay connected.

    Bug: 110642515
    Test: Manual test, make a call and switch path A->speaker->B on phone UI

    Change-Id: I7dde255c2ba9f04aec86b7fdd30e7bfadfdd5068
    (cherry picked from commit aebfbb7e35ca4cafaa9167d74a37fdb9bc8f62bc)

commit bd3055fe974818637c85c37aa3305cd6ec70f6a7
Author: Cheney Ni <cheneyni@google.com>
Date:   Thu Jun 21 20:19:03 2018 +0800

    Always assign a free ID to the new MediaPlayer

    It used current size plus 1 as the new MediaPlayer's ID in previous
    design and if a player was removed before, the ID of size+1 maybe was
    still in used but a ID < size+1 was freed.

    This change finds a free ID by checking if the number is not in the
    mMediaPlayerIds so we never use same ID in two player.

    Bug: 110511529
    Test: manually switching players by force-stop / disabled
    runtest bluetooth -c com.android.bluetooth.avrcp.MediaPlayerWrapperTest

    Change-Id: Ia9e67a34b5915ad615746e6b334c1596ee639c1a
    (cherry picked from commit e2da76c960a90d9d9e52ef2cc92a5e7a00c122d1)

commit 3bb8f7841c7427a3dc2d0217ca7482bb8b0296fe
Author: Jack He <siyuanh@google.com>
Date:   Thu Jun 14 19:30:02 2018 -0700

    HFP: Use SignalStrength#getLevel() to get signal level

    * According to TS 127 007 v6.8.0 Section 8.9, +CIND indicator value for
      "signal" should be 0 when "the indicator is off (or in state which can
      be identified as 'off'-state)". Therefore, this CL only assigns +CIND
      "signal" indicator as 0 when cellular network is unavailable
    * When network state is available, call SignalStrength#getLevel() to get
      network signal level in the range of 0 to 4 and add 1 to this level to
      get +CIND signal level

    Bug: 109857475
    Test: make, connect to car kit at different network scenario
    Change-Id: I9c737813fd779351efc387ca14eb41b56d6e02e2
    (cherry picked from commit 340b4aa1e07198db3d4e577267865d09a670e64c)

commit 51a648ef846e94f4a33596dd72ee11295b4ecc93
Merge: b9319866e 5851f9b5a
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Jun 13 03:09:05 2018 +0000

    Snap for 4836998 from 5851f9b5afe30aea8625e56fc2192befb63c663b to pi-dr1-release

    Change-Id: I5f6f9c2dd193206df6a11f426ef03ca8a0d3b764

commit 5851f9b5afe30aea8625e56fc2192befb63c663b
Author: Ajay Panicker <apanicke@google.com>
Date:   Tue Jun 12 10:34:39 2018 -0700

    Prevent crash when a Browsable player has an empty root

    Bug: 110051940
    Test: runtest bluetooth -c
    com.android.bluetooth.avrcp.BrowserPlayerWrapperTest

    Change-Id: I2d6db2814f3ea61e95e0a1cc8417274f103af5de
    (cherry picked from commit acdf7d6bc28bdda01a3918772d0082fa75bc91c5)

commit b9319866e01d4891ca8c04a7aa9d084bea9af9fd
Merge: f671214a6 1f5d15066
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Tue Jun 12 03:06:08 2018 +0000

    Snap for 4834273 from 1f5d1506692310e7ceaf5959058a849ee5711cf9 to pi-dr1-release

    Change-Id: I87c827fd0ea7b104184fa5cf191923d2a74e122d

commit f671214a61c49fd1e64fac4c2eb6e476ad639b70
Merge: 790bcaa3a e655d4f8e
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Sun Jun 10 03:05:37 2018 +0000

    Snap for 4832091 from e655d4f8e730346c7f424bd9ddbb7c2a6fd90648 to pi-dr1-release

    Change-Id: If2fc3099db7f2cbaf9986f65cb172ffb407a47b0

commit 1f5d1506692310e7ceaf5959058a849ee5711cf9
Author: Bill Yi <byi@google.com>
Date:   Sat Jun 9 07:18:10 2018 -0700

    Import translations. DO NOT MERGE

    Auto-generated-cl: translation import

    Bug: 64712476
    Change-Id: I8a8f6e0785174da53e57c694507bfdc9e86328fe

commit 790bcaa3a96fa0dc569eb331e999ce5dfb08af41
Merge: 10701c185 7760d083a
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Jun 6 03:04:57 2018 +0000

    Snap for 4823496 from 7760d083a490110a4d0ac64a9e4689cebbd12491 to pi-dr1-release

    Change-Id: Ib661d02dfd2651777f04efcd96a9033b1f82870f

commit 10701c185fdd095403a54a74d5fe64d1654edcef
Merge: af6d94db8 e1aa31860
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Tue Jun 5 03:02:45 2018 +0000

    Snap for 4820872 from e1aa31860e1dfbedff375560e6d979497aef750f to pi-dr1-release

    Change-Id: I2fa594b53b7bb9387341c15635e63539b110db1f

Change-Id: I6d2c0c7a23447d2f1b799ed16ac7f68f15818d6f
---
 res/values-ar/strings.xml                     |   2 +-
 res/values-fa/strings_pbap.xml                |   4 +-
 res/values-pl/strings.xml                     |   2 +-
 res/values-pt/strings.xml                     |   2 +-
 .../android/bluetooth/a2dp/A2dpService.java   |  11 ++
 .../bluetooth/hfp/HeadsetPhoneState.java      | 122 +-----------------
 .../android/bluetooth/hfp/HeadsetService.java |   6 +-
 .../bluetooth/hfp/HeadsetStateMachine.java    |   3 +-
 .../bluetooth/newavrcp/MediaPlayerList.java   |  10 +-
 9 files changed, 32 insertions(+), 130 deletions(-)

diff --git a/res/values-ar/strings.xml b/res/values-ar/strings.xml
index bb6db58c..4bb0f8a0 100644
--- a/res/values-ar/strings.xml
+++ b/res/values-ar/strings.xml
@@ -26,7 +26,7 @@
     <string name="airplane_error_title" msgid="2683839635115739939">"وضع الطائرة"</string>
     <string name="airplane_error_msg" msgid="8698965595254137230">"لا يمكنك استخدام البلوتوث في وضع الطائرة."</string>
     <string name="bt_enable_title" msgid="8657832550503456572"></string>
-    <string name="bt_enable_line1" msgid="7203551583048149">"إلى لاستخدم خدمات البلوتوث، يجب أولاً تشغيل البلوتوث."</string>
+    <string name="bt_enable_line1" msgid="7203551583048149">"لإستخدام خدمات البلوتوث، يجب تشغيل البلوتوث أولاً."</string>
     <string name="bt_enable_line2" msgid="4341936569415937994">"هل تريد تشغيل البلوتوث الآن؟"</string>
     <string name="bt_enable_cancel" msgid="1988832367505151727">"إلغاء"</string>
     <string name="bt_enable_ok" msgid="3432462749994538265">"تشغيل"</string>
diff --git a/res/values-fa/strings_pbap.xml b/res/values-fa/strings_pbap.xml
index 8282443d..348156a6 100644
--- a/res/values-fa/strings_pbap.xml
+++ b/res/values-fa/strings_pbap.xml
@@ -3,8 +3,8 @@
     xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
     <string name="pbap_session_key_dialog_title" msgid="3580996574333882561">"‏تایپ کلید جلسه برای %1$s"</string>
     <string name="pbap_session_key_dialog_header" msgid="2772472422782758981">"به کلید جلسه بلوتوث نیاز است"</string>
-    <string name="pbap_acceptance_timeout_message" msgid="1107401415099814293">"‏هنگام پذیرش اتصال به %1$s وقفه زمانی ایجاد شده است"</string>
-    <string name="pbap_authentication_timeout_message" msgid="4166979525521902687">"‏هنگام ورود کلید جلسه با %1$s وقفه زمانی ایجاد شده است"</string>
+    <string name="pbap_acceptance_timeout_message" msgid="1107401415099814293">"‏پذیرش اتصال با %1$s خیلی زمان برد"</string>
+    <string name="pbap_authentication_timeout_message" msgid="4166979525521902687">"‏وارد کردن کلید جلسه با %1$s خیلی زمان برد"</string>
     <string name="auth_notif_ticker" msgid="1575825798053163744">"‏درخواست احراز هویت Obex"</string>
     <string name="auth_notif_title" msgid="7599854855681573258">"کلید جلسه"</string>
     <string name="auth_notif_message" msgid="6667218116427605038">"‏تایپ کلید جلسه برای %1$s"</string>
diff --git a/res/values-pl/strings.xml b/res/values-pl/strings.xml
index f1b4575b..ee096509 100644
--- a/res/values-pl/strings.xml
+++ b/res/values-pl/strings.xml
@@ -78,7 +78,7 @@
     <string name="not_exist_file_desc" msgid="4059531573790529229">"Plik nie istnieje. \n"</string>
     <string name="enabling_progress_title" msgid="436157952334723406">"Czekaj…"</string>
     <string name="enabling_progress_content" msgid="4601542238119927904">"Włączanie Bluetooth…"</string>
-    <string name="bt_toast_1" msgid="972182708034353383">"Plik zostanie odebrany. Sprawdzaj postęp na panelu powiadomień."</string>
+    <string name="bt_toast_1" msgid="972182708034353383">"Plik zostanie odebrany. Sprawdzaj postęp w panelu powiadomień."</string>
     <string name="bt_toast_2" msgid="8602553334099066582">"Nie można odebrać pliku."</string>
     <string name="bt_toast_3" msgid="6707884165086862518">"Zatrzymano odbiór pliku z urządzenia „<xliff:g id="SENDER">%1$s</xliff:g>”"</string>
     <string name="bt_toast_4" msgid="4678812947604395649">"Wysyłanie pliku do urządzenia „<xliff:g id="RECIPIENT">%1$s</xliff:g>”"</string>
diff --git a/res/values-pt/strings.xml b/res/values-pt/strings.xml
index 9b4d14fa..d024a424 100644
--- a/res/values-pt/strings.xml
+++ b/res/values-pt/strings.xml
@@ -124,7 +124,7 @@
     <string name="transfer_clear_dlg_title" msgid="2953444575556460386">"Limpar"</string>
     <string name="bluetooth_map_settings_save" msgid="7635491847388074606">"Salvar"</string>
     <string name="bluetooth_map_settings_cancel" msgid="9205350798049865699">"Cancelar"</string>
-    <string name="bluetooth_map_settings_intro" msgid="6482369468223987562">"Selecione as contas que você deseja compartilhar via Bluetooth. Você ainda precisa aceitar qualquer acesso às contas ao se conectar."</string>
+    <string name="bluetooth_map_settings_intro" msgid="6482369468223987562">"Selecione as contas que você quer compartilhar via Bluetooth. Você ainda precisa aceitar qualquer acesso às contas ao se conectar."</string>
     <string name="bluetooth_map_settings_count" msgid="4557473074937024833">"Espaços disponíveis:"</string>
     <string name="bluetooth_map_settings_app_icon" msgid="7105805610929114707">"Ícone do app"</string>
     <string name="bluetooth_map_settings_title" msgid="7420332483392851321">"Configurações de compartilhamento de mensagens Bluetooth"</string>
diff --git a/src/com/android/bluetooth/a2dp/A2dpService.java b/src/com/android/bluetooth/a2dp/A2dpService.java
index 6ba95f3a..361ac673 100644
--- a/src/com/android/bluetooth/a2dp/A2dpService.java
+++ b/src/com/android/bluetooth/a2dp/A2dpService.java
@@ -517,7 +517,14 @@ public class A2dpService extends ProfileService {
                 }
                 // Make sure the Audio Manager knows the previous Active device is disconnected,
                 // and the new Active device is connected.
+                // Also, mute and unmute the output during the switch to avoid audio glitches.
+                boolean wasMuted = false;
                 if (previousActiveDevice != null) {
+                    if (!mAudioManager.isStreamMute(AudioManager.STREAM_MUSIC)) {
+                        mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC,
+                                                         AudioManager.ADJUST_MUTE, 0);
+                        wasMuted = true;
+                    }
                     mAudioManager.setBluetoothA2dpDeviceConnectionStateSuppressNoisyIntent(
                             previousActiveDevice, BluetoothProfile.STATE_DISCONNECTED,
                             BluetoothProfile.A2DP, true, -1);
@@ -539,6 +546,10 @@ public class A2dpService extends ProfileService {
                 // change, so the Audio Service can reset accordingly the audio
                 // feeding parameters in the Audio HAL to the Bluetooth stack.
                 mAudioManager.handleBluetoothA2dpDeviceConfigChange(mActiveDevice);
+                if (wasMuted) {
+                    mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC,
+                                                     AudioManager.ADJUST_UNMUTE, 0);
+                }
             }
         }
         return true;
diff --git a/src/com/android/bluetooth/hfp/HeadsetPhoneState.java b/src/com/android/bluetooth/hfp/HeadsetPhoneState.java
index a0914384..bcf123cd 100644
--- a/src/com/android/bluetooth/hfp/HeadsetPhoneState.java
+++ b/src/com/android/bluetooth/hfp/HeadsetPhoneState.java
@@ -323,134 +323,18 @@ public class HeadsetPhoneState {
 
         @Override
         public void onSignalStrengthsChanged(SignalStrength signalStrength) {
-
             int prevSignal = mCindSignal;
             if (mCindService == HeadsetHalConstants.NETWORK_STATE_NOT_AVAILABLE) {
                 mCindSignal = 0;
-            } else if (signalStrength.isGsm()) {
-                mCindSignal = signalStrength.getLteLevel();
-                if (mCindSignal == SignalStrength.SIGNAL_STRENGTH_NONE_OR_UNKNOWN) {
-                    mCindSignal = gsmAsuToSignal(signalStrength);
-                } else {
-                    // SignalStrength#getLteLevel returns the scale from 0-4
-                    // Bluetooth signal scales at 0-5
-                    // Let's match up the larger side
-                    mCindSignal++;
-                }
             } else {
-                mCindSignal = cdmaDbmEcioToSignal(signalStrength);
+                mCindSignal = signalStrength.getLevel() + 1;
             }
-
-            // network signal strength is scaled to BT 1-5 levels.
+            // +CIND "signal" indicator is always between 0 to 5
+            mCindSignal = Integer.max(Integer.min(mCindSignal, 5), 0);
             // This results in a lot of duplicate messages, hence this check
             if (prevSignal != mCindSignal) {
                 sendDeviceStateChanged();
             }
         }
-
-        /* convert [0,31] ASU signal strength to the [0,5] expected by
-         * bluetooth devices. Scale is similar to status bar policy
-         */
-        private int gsmAsuToSignal(SignalStrength signalStrength) {
-            int asu = signalStrength.getGsmSignalStrength();
-            if (asu == 99) {
-                return 0;
-            } else if (asu >= 16) {
-                return 5;
-            } else if (asu >= 8) {
-                return 4;
-            } else if (asu >= 4) {
-                return 3;
-            } else if (asu >= 2) {
-                return 2;
-            } else if (asu >= 1) {
-                return 1;
-            } else {
-                return 0;
-            }
-        }
-
-        /**
-         * Convert the cdma / evdo db levels to appropriate icon level.
-         * The scale is similar to the one used in status bar policy.
-         *
-         * @param signalStrength signal strength level
-         * @return the icon level for remote device
-         */
-        private int cdmaDbmEcioToSignal(SignalStrength signalStrength) {
-            int levelDbm = 0;
-            int levelEcio = 0;
-            int cdmaIconLevel = 0;
-            int evdoIconLevel = 0;
-            int cdmaDbm = signalStrength.getCdmaDbm();
-            int cdmaEcio = signalStrength.getCdmaEcio();
-
-            if (cdmaDbm >= -75) {
-                levelDbm = 4;
-            } else if (cdmaDbm >= -85) {
-                levelDbm = 3;
-            } else if (cdmaDbm >= -95) {
-                levelDbm = 2;
-            } else if (cdmaDbm >= -100) {
-                levelDbm = 1;
-            } else {
-                levelDbm = 0;
-            }
-
-            // Ec/Io are in dB*10
-            if (cdmaEcio >= -90) {
-                levelEcio = 4;
-            } else if (cdmaEcio >= -110) {
-                levelEcio = 3;
-            } else if (cdmaEcio >= -130) {
-                levelEcio = 2;
-            } else if (cdmaEcio >= -150) {
-                levelEcio = 1;
-            } else {
-                levelEcio = 0;
-            }
-
-            cdmaIconLevel = (levelDbm < levelEcio) ? levelDbm : levelEcio;
-
-            // STOPSHIP: Change back to getRilVoiceRadioTechnology
-            if (mServiceState != null && (
-                    mServiceState.getRadioTechnology() == ServiceState.RIL_RADIO_TECHNOLOGY_EVDO_0
-                            || mServiceState.getRadioTechnology()
-                            == ServiceState.RIL_RADIO_TECHNOLOGY_EVDO_A)) {
-                int evdoEcio = signalStrength.getEvdoEcio();
-                int evdoSnr = signalStrength.getEvdoSnr();
-                int levelEvdoEcio = 0;
-                int levelEvdoSnr = 0;
-
-                // Ec/Io are in dB*10
-                if (evdoEcio >= -650) {
-                    levelEvdoEcio = 4;
-                } else if (evdoEcio >= -750) {
-                    levelEvdoEcio = 3;
-                } else if (evdoEcio >= -900) {
-                    levelEvdoEcio = 2;
-                } else if (evdoEcio >= -1050) {
-                    levelEvdoEcio = 1;
-                } else {
-                    levelEvdoEcio = 0;
-                }
-
-                if (evdoSnr > 7) {
-                    levelEvdoSnr = 4;
-                } else if (evdoSnr > 5) {
-                    levelEvdoSnr = 3;
-                } else if (evdoSnr > 3) {
-                    levelEvdoSnr = 2;
-                } else if (evdoSnr > 1) {
-                    levelEvdoSnr = 1;
-                } else {
-                    levelEvdoSnr = 0;
-                }
-
-                evdoIconLevel = (levelEvdoEcio < levelEvdoSnr) ? levelEvdoEcio : levelEvdoSnr;
-            }
-            // TODO(): There is a bug open regarding what should be sent.
-            return (cdmaIconLevel > evdoIconLevel) ? cdmaIconLevel : evdoIconLevel;
-        }
     }
 }
diff --git a/src/com/android/bluetooth/hfp/HeadsetService.java b/src/com/android/bluetooth/hfp/HeadsetService.java
index db115ca2..2ccc1e48 100644
--- a/src/com/android/bluetooth/hfp/HeadsetService.java
+++ b/src/com/android/bluetooth/hfp/HeadsetService.java
@@ -1486,13 +1486,13 @@ public class HeadsetService extends ProfileService {
             }
         }
         mStateMachinesThread.getThreadHandler().post(() -> {
-            boolean shouldCallAudioBeActiveBefore = shouldCallAudioBeActive();
+            boolean isCallIdleBefore = mSystemInterface.isCallIdle();
             mSystemInterface.getHeadsetPhoneState().setNumActiveCall(numActive);
             mSystemInterface.getHeadsetPhoneState().setNumHeldCall(numHeld);
             mSystemInterface.getHeadsetPhoneState().setCallState(callState);
             // Suspend A2DP when call about is about to become active
             if (callState != HeadsetHalConstants.CALL_STATE_DISCONNECTED
-                    && shouldCallAudioBeActive() && !shouldCallAudioBeActiveBefore) {
+                    && !mSystemInterface.isCallIdle() && isCallIdleBefore) {
                 mSystemInterface.getAudioManager().setParameters("A2dpSuspended=true");
             }
         });
@@ -1501,7 +1501,7 @@ public class HeadsetService extends ProfileService {
                         new HeadsetCallState(numActive, numHeld, callState, number, type)));
         mStateMachinesThread.getThreadHandler().post(() -> {
             if (callState == HeadsetHalConstants.CALL_STATE_IDLE
-                    && !shouldCallAudioBeActive() && !isAudioOn()) {
+                    && mSystemInterface.isCallIdle() && !isAudioOn()) {
                 // Resume A2DP when call ended and SCO is not connected
                 mSystemInterface.getAudioManager().setParameters("A2dpSuspended=false");
             }
diff --git a/src/com/android/bluetooth/hfp/HeadsetStateMachine.java b/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
index b1d2cfc7..d25d9edf 100644
--- a/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
+++ b/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
@@ -1201,7 +1201,8 @@ public class HeadsetStateMachine extends StateMachine {
             // Set active device to current active SCO device when the current active device
             // is different from mCurrentDevice. This is to accommodate active device state
             // mis-match between native and Java.
-            if (!mDevice.equals(mHeadsetService.getActiveDevice())) {
+            if (!mDevice.equals(mHeadsetService.getActiveDevice())
+                    && !hasDeferredMessages(DISCONNECT_AUDIO)) {
                 mHeadsetService.setActiveDevice(mDevice);
             }
             setAudioParameters();
diff --git a/src/com/android/bluetooth/newavrcp/MediaPlayerList.java b/src/com/android/bluetooth/newavrcp/MediaPlayerList.java
index 1a10979c..74ea0393 100644
--- a/src/com/android/bluetooth/newavrcp/MediaPlayerList.java
+++ b/src/com/android/bluetooth/newavrcp/MediaPlayerList.java
@@ -152,7 +152,7 @@ public class MediaPlayerList {
                 for (BrowsedPlayerWrapper wrapper : players) {
                     // Generate new id and add the browsable player
                     if (!mMediaPlayerIds.containsKey(wrapper.getPackageName())) {
-                        mMediaPlayerIds.put(wrapper.getPackageName(), mMediaPlayerIds.size() + 1);
+                        mMediaPlayerIds.put(wrapper.getPackageName(), getFreeMediaPlayerId());
                     }
 
                     d("Adding Browser Wrapper for " + wrapper.getPackageName() + " with id "
@@ -206,6 +206,12 @@ public class MediaPlayerList {
         return BLUETOOTH_PLAYER_ID;
     }
 
+    int getFreeMediaPlayerId() {
+        int id = 0;
+        while (mMediaPlayerIds.containsValue(++id)) {}
+        return id;
+    }
+
     MediaPlayerWrapper getActivePlayer() {
         return mMediaPlayers.get(mActivePlayerId);
     }
@@ -405,7 +411,7 @@ public class MediaPlayerList {
         // that key.
         String packageName = controller.getPackageName();
         if (!mMediaPlayerIds.containsKey(packageName)) {
-            mMediaPlayerIds.put(packageName, mMediaPlayerIds.size() + 1);
+            mMediaPlayerIds.put(packageName, getFreeMediaPlayerId());
         }
 
         int playerId = mMediaPlayerIds.get(packageName);
-- 
2.17.1

